Whitelist URL class

interpreter95 opened the pull request:
Exposes parts of the [java.net.URL](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html) class in painless.

API:
```
class java.net.URL {
    (String)
    String getHost()
    String getPath()
    int getPort()
    String getProtocol()
    String getRef()
}
```

Closes: #60660

interpreter95 committed:
Whitelist URL class

Exposes parts of the [java.net.URL](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html) class in painless.

API:
```
class java.net.URL {
    (String)
    String getHost()
    String getPath()
    int getPort()
    String getProtocol()
    String getRef()
}
```

Closes: #60660

interpreter95 commented:
I just want to make the reviewer aware that when I ran `./gradlew check` locally I seem to fail `:libs:elasticsearch-ssl-config:test` with error message `elasticsearch Cannot find trusted issuer with name [thawte]`. I assume that this failure is because of a configuration error and not part of my commit. If not, please let me know and I will debug further!

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

stu-elastic commented:
@elasticmachine ok to test

stu-elastic commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/UrlTests.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public void testGetHost() {
+        assertEquals("www.test.com", exec("URL url = new URL('https://www.test.com:8080/path?query#fragment'); url.getHost()"));
```
Please use `example.com`

interpreter95 replied:
Will do!

stu-elastic commented:
We're going to have to override `hashCode()` and `equals()` because they are blocking operations.

interpreter95 committed:
Fix: Change URLs from test.com to example.com

As requested, the URLs used in the testing of
the whitelisting of the java.net.URL class
should be on the form "example.com" instead
of "test.com".

interpreter95 commented:
Hi,

I see in the Jenkins logs that my pull request is failing tests because of:

```
15:21:49 runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
15:21:49 runbld>>> Adding /var/lib/jenkins/.java/openjdk-15.0.2+7-linux/bin to the path.
15:21:49 + /dev/shm/elastic+elasticsearch+master+pull-request+bwc/.ci/scripts/run-gradle.sh -Dignore.tests.seed bwcTestSnapshots
15:21:49 /tmp/jenkins1340635122548236278.sh: line 2: /dev/shm/elastic+elasticsearch+master+pull-request+bwc/.ci/scripts/run-gradle.sh: No such file or directory
15:21:49 runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
```

Seems like a file or folder has been moved... But from what I know, I have only modified/added files in my pull request.

Anyone have any hints on what I should do to fix this?

rjernst commented:
> Anyone have any hints on what I should do to fix this?

Our CI infrastructure changed. You'll need to sync with the latest from the `master` branch.

stu-elastic commented:
@interpreter95 We're going to have to figure out a way to automatically specify a [URLStreamHandler](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#URL(java.net.URL,%20java.lang.String,%20java.net.URLStreamHandler)) that performs no-ops. 

Otherwise the [system](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#URL(java.lang.String,%20java.lang.String,%20int,%20java.lang.String)) `URLStreamHandler` is used for the given protocol.  We can't have `equals` and `hashcode` be blocking network operations and a dummy URLStreamHandler is the best way to avoid any surprises.

After we've solved that issue, we can modify this PR to get it merged.

interpreter95 committed:
Fix: Remove error message check for 'null' case

It has been observed that when supplying 'null' to the
constructor of java.net.URL we get an error with an error message
in Java 15 but in Java 11 we get an error without an error message.

Thus, the check for the error message has been removed from testing
the 'null' scenario, as depending on the Java version used during testing
the check might either pass or fail.

Instead the error message check has been replaced with a check that
the error thrown is not 'null'.

interpreter95 commented:
Thanks for the update @stu-elastic 😃 

I'm just cleaning up the code a bit, and also fixing a bug in my tests that the latest test of the PR found!

Feel free to update this PR as you like, I just wanted to leave it in a state where it's hopefully easy for you to modify it.

I will submit an update to the PR today just to check that all the tests pass. Then I leave it up to you to decide what to do with the PR.

stu-elastic commented:
Hi @interpreter95, as mentioned in https://github.com/elastic/elasticsearch/issues/60660#issuecomment-875808518, we're going to go another way for this task.  This is the correct way to whitelist and the correct way to test whitelisted methods.  I hope you'll pick another painless issue to work on.  Feel free to ping me on whichever issue you pick up next and I'll endeavor to be more responsive.

Thanks.
